kotlinx-coroutines-test/common/src/TestBuilders.kt

```diff
@@ -269,46 +278,126 @@ public fun runTest(
  *
  * @throws IllegalArgumentException if the [context] is invalid. See the [TestScope] constructor docs for details.
  */
-@ExperimentalCoroutinesApi
+@Deprecated(
+    "Define a total timeout for the whole test instead of using dispatchTimeoutMs. " +
+        "Warning: the proposed replacement is not identical as it uses 'dispatchTimeoutMs' as the timeout for the whole test!",
+    ReplaceWith("runTest(context, timeout = dispatchTimeoutMs.milliseconds, testBody)",
+        "kotlin.time.Duration.Companion.milliseconds"),
+    DeprecationLevel.WARNING
+)
 public fun runTest(
     context: CoroutineContext = EmptyCoroutineContext,
     dispatchTimeoutMs: Long,
     testBody: suspend TestScope.() -> Unit
-): TestResult = runTest(
-    context = context,
-    dispatchTimeout = dispatchTimeoutMs.milliseconds,
-    testBody = testBody
-)
+): TestResult {
+    if (context[RunningInRunTest] != null)
+        throw IllegalStateException("Calls to `runTest` can't be nested. Please read the docs on `TestResult` for details.")
+    @Suppress("DEPRECATION")
+    return TestScope(context + RunningInRunTest).runTest(dispatchTimeoutMs = dispatchTimeoutMs, testBody)
+}
 
 /**
- * Performs [runTest] on an existing [TestScope].
+ * Performs [runTest] on an existing [TestScope]. See the documentation for [runTest] for details.
  */
 @ExperimentalCoroutinesApi
 public fun TestScope.runTest(
-    dispatchTimeout: Duration,
+    timeout: Duration = DEFAULT_TIMEOUT,
     testBody: suspend TestScope.() -> Unit
 ): TestResult = asSpecificImplementation().let {
     it.enter()
     createTestResult {
-        runTestCoroutine(it, dispatchTimeout, TestScopeImpl::tryGetCompletionCause, testBody) {
+        /** TODO: moving this [AbstractCoroutine.start] call outside [createTestResult] fails on JS. */
+        it.start(CoroutineStart.UNDISPATCHED, it) {
+            testBody()
+        }
+        /**
```

Comment:
I don't think we should be concerned about the performance of failing tests unless we're talking orders of magnitude, as failing tests are already outliers. Tests that complete before the first suspension are much more likely: for example, unit tests of the form `runTest { something.viewModelScope.launch { ... }; assertSomething() }` finish without suspensions.

We should check the performance impact of the code you're proposing on some actual code base. I could easily believe it if someone told me that the gain from this is dwarfed by just the overhead of starting a new test. A 20% improvement in total runtime for a common class of unit tests would be great, but if the improvement is something like 0.5%, I wouldn't want to introduce such code. If you want, I will elaborate on the many reasons, but the gist is that it's one more edge case to think about.

Reply:
The reasoning works for me, thanks for the explanation.

>but the gist is that it's one more edge case to think about.

I've noticed that all the machinery that follows `testBody()` does [if the test is completed] nothing except for creating system-level jobs/coroutines and immediately completing them. 
I doubt there will be any _significant_ improvement (because the condition covers a really specific subset of tests: failed, not suspended, and `assume`-style ones), and was tempted to think about it as "minor improvement that costs nothing to add". Your concern about this condition being harder to reason about is enough for me to discard the idea.

Reply:
Well, the change is not applicable with the latest revision anyway, as the test body never completes in the undispatched invocation, so not even relevant anymore.